### DISCUSSION TITLE
course/starbook/session-get-user-info

### DISCUSSION THREAD

**Discussion by giscus** [gogf/gf-site#114] [General]
# course/starbook/session-get-user-info

用户信息接口需要登录后才能访问，并通过中间件统一验证Token的有效性。GoFrame框架提供了灵活的中间件机制，支持请求前后操作。通过Auth中间件验证HTTP请求中的Token，在Logic中解析Token获取用户信息，并通过定义Api和Controller调用控制器进行接口注册，实现用户信息接口的访问控制和权限认证。

https://goframe.org/course/starbook/session-get-user-info

**Comment by Abner3721** [comment]
这个少步骤了，缺少account_new.go文件，和star/internal/controller/account
/account.go文件

**Reply by Abner3721** [reply]
还缺少api/account/account.go这个文件

**Reply by oldme-git** [reply]
这两个文件是自动生成的，用到才需编写

**Comment by walkmiao** [comment]
为什么我这个g.RequestFromCtx(ctx).Request 为 nil 和上面代码一致

**Reply by NiceLisan** [reply]
是的 我也测试了半天 g.RequestFromCtx(ctx).Request 为 nil 有人解决了么。
已经改过  jwtKey = []byte(consts.JwtKey) // 确保密钥是 []byte 类型，好像g.RequestFromCtx(ctx).Request 一直是nil，
这里被卡住了

**Comment by coinpwd** [comment]
"star/utility"中并没有设置 jwtKey，配置了jwtkey的路径在	"star/internal/consts"

jwtKey      = utility.JwtKey

**Comment by coinpwd** [comment]
token signature is invalid: key is of invalid type: HMAC verify expects []byte
需要改		jwtKey      = []byte(consts.JwtKey) // 确保密钥是 []byte 类型

**Comment by ichenyy** [comment]
这句代码需要调整一下：

``` go
token, err := jwt.Parse(tokenString, func(token *jwt.Token) (interface{}, error) {  
       return jwtKey, nil  
}) 
```

改成：

```go
token, err := jwt.Parse(tokenString, func(token *jwt.Token) (interface{}, error) {  
       return []byte(jwtKey), nil  
})  
```

**Reply by Abner3721** [reply]
这个教程咋各种问题，跟着教程都走不通。

**Comment by Abner3721** [comment]
internal/logic/middleware/auth.go
这个文件的return consts.JwtKey, nil 地方应该改成切片的形式如：return []byte(consts.JwtKey), nil
整个文件代码如下：

package middleware

import (
	"net/http"

	"star/internal/consts"

	"github.com/gogf/gf/v2/net/ghttp"
	"github.com/golang-jwt/jwt/v5"
)

func Auth(r *ghttp.Request) {
	var tokenString = r.Header.Get("Authorization")
	token, err := jwt.Parse(tokenString, func(token *jwt.Token) (interface{}, error) {
		return []byte(consts.JwtKey), nil
	})
	if err != nil || !token.Valid {
		r.Response.WriteStatus(http.StatusForbidden)
		r.Exit()
	}
	r.Middleware.Next()
}






/internal/logic/users/users_account.go

这个文件里面的代码consts.JwtKey 应该改成 切片的形式如：[]byte(consts.JwtKey)
整个文件的代码如下：
package users

import (
	"context"
	"time"

	"star/internal/consts"
	"star/internal/dao"
	"star/internal/model/entity"

	"github.com/gogf/gf/v2/errors/gerror"
	"github.com/gogf/gf/v2/frame/g"
	"github.com/golang-jwt/jwt/v5"
)

type jwtClaims struct {
	Id       uint
	Username string
	jwt.RegisteredClaims
}

func (u *Users) Login(ctx context.Context, username, password string) (tokenString string, err error) {
	var user entity.Users
	err = dao.Users.Ctx(ctx).Where("username", username).Scan(&user)
	if err != nil {
		return "", gerror.New("用户名或密码错误")
	}

	if user.Id == 0 {
		return "", gerror.New("用户不存在")
	}

	// 将密码加密后与数据库中的密码进行比对
	if user.Password != u.encryptPassword(password) {
		return "", gerror.New("用户名或密码错误")
	}

	// 生成token
	uc := &jwtClaims{
		Id:       user.Id,
		Username: user.Username,
		RegisteredClaims: jwt.RegisteredClaims{
			ExpiresAt: jwt.NewNumericDate(time.Now().Add(6 * time.Hour)),
		},
	}
	token := jwt.NewWithClaims(jwt.SigningMethodHS256, uc)
	return token.SignedString([]byte(consts.JwtKey))
}

func (u *Users) Info(ctx context.Context) (user *entity.Users, err error) {
	tokenString := g.RequestFromCtx(ctx).Request.Header.Get("Authorization")
	tokenClaims, _ := jwt.ParseWithClaims(tokenString, &jwtClaims{}, func(token *jwt.Token) (interface{}, error) {
		return []byte(consts.JwtKey), nil
	})

	if claims, ok := tokenClaims.Claims.(*jwtClaims); ok && tokenClaims.Valid {
		err = dao.Users.Ctx(ctx).Where("id", claims.Id).Scan(&user)
	}
	return

}

**Reply by xjsadmin** [reply]
不错，按照你的方法可以正常运行的。

**Comment by hqiaozhi** [comment]
所有使用consts.JwtKey的地方都要改成[]byte(consts.JwtKey),不然就会各种各样报错，大家都踩坑了，哈哈哈~

**Comment by Freya789** [comment]
为什么从http://127.0.0.1:8000/v1/account/info获取到"Forbidden", 而不是用户信息？我打开这个网址也只有这一个字符串。

**Reply by 1ZYHLOG5BI** [reply]
可以看看前面老哥说的，试试把consts.JwtKey的地方都改成[]byte(consts.JwtKey)

**Reply by tonyho** [reply]
增加了 api/account/v1/account.go 后是不是需要 `gf gen ctrl` ?

**Reply by NiceLisan** [reply]
这是因为g.RequestFromCtx(ctx).Request 一直返回nil,
不知道有谁可以修复这个错误？

**Comment by NiceLisan** [comment]
auth.go 需要做如下改动：
func Auth(r *ghttp.Request) {
	var tokenString = r.Header.Get("Authorization")

	g.Log().Info("token Authorization: " + tokenString)

	token, err := jwt.Parse(tokenString, func(token *jwt.Token) (interface{}, error) {
		return []byte(consts.JwtKey), nil
	})
	if err != nil || !token.Valid {

		r.Response.WriteStatus(http.StatusForbidden)

		r.Exit()
	}
	ctx := r.Context()
	ctx = context.WithValue(ctx, "Authorization", tokenString)

	r.SetCtx(ctx)
	r.Middleware.Next()
}
user_account.go的Info函数需要做如下改动：
func (u *Users) Info(ctx context.Context) (user *entity.Users, err error) {

	tokenString := ctx.Value("Authorization").(string)
	g.Log().Info("user_account token: " + tokenString)
	tokenClaims, _ := jwt.ParseWithClaims(tokenString, &jwtClaims{}, func(token *jwt.Token) (interface{}, error) {
		return []byte(consts.JwtKey), nil
	})

	if claims, ok := tokenClaims.Claims.(*jwtClaims); ok && tokenClaims.Valid {
		err = dao.Users.Ctx(ctx).Where("id", claims.Id).Scan(&user)
	}
	return
}
这样从http://127.0.0.1:8000/v1/account/info 就可以获取到已经登陆用户的信息了。
